src/iface/interface.rs

```diff
@@ -1830,22 +1908,105 @@ impl<'a> InterfaceInner<'a> {
         &mut self,
         sockets: &mut SocketSet,
-        ipv4_packet: &Ipv4Packet<&'frame T>,
-    ) -> Option<IpPacket<'frame>> {
+        ipv4_packet: &Ipv4Packet<&'payload T>,
+        _fragments: Option<&'output mut PacketAssemblerSet<'a, Ipv4FragKey>>,
+    ) -> Option<IpPacket<'output>> {
         let ipv4_repr = check!(Ipv4Repr::parse(ipv4_packet, &self.caps.checksum));
-
         if !self.is_unicast_v4(ipv4_repr.src_addr) {
             // Discard packets with non-unicast source addresses.
             net_debug!("non-unicast source address");
             return None;
         }
 
-        let ip_repr = IpRepr::Ipv4(ipv4_repr);
+        #[cfg(feature = "proto-ipv4-fragmentation")]
+        let ip_payload = {
+            const REASSEMBLY_TIMEOUT: u64 = 90;
+
+            let fragments = _fragments.unwrap();
+
+            check!(fragments
+                .mark_discarded_when(|frag| Ok(self.now - frag.start_time()? > frag.ttl()?)));
+
+            if ipv4_packet.more_frags() || ipv4_packet.frag_offset() != 0 {
+                let key = ipv4_packet.get_key();
+
+                if let Ok(f) = fragments.get_packet_assembler_mut(&key) {
+                    if f.is_discarded() {
+                        fragments.remove_discarded();
+
+                        // FIXME: we should send the 8 first octets of fragment 0.
+                        let icmp_reply_repr = Icmpv4Repr::TimeExceeded {
+                            reason: Icmpv4TimeExceeded::FragExpired,
+                            header: ipv4_repr,
+                            data: ipv4_packet.payload(),
+                        };
+
+                        return self.icmpv4_reply(ipv4_repr, icmp_reply_repr);
+                    }
+                };
+
+                fragments.remove_discarded();
+
+                let f = match fragments.get_packet_assembler_mut(&key) {
+                    Ok(f) => f,
+                    Err(_) => {
+                        check!(check!(fragments.reserve_with_key(&key)).start(
+                            None,
+                            self.now,
+                            Duration::from_secs(REASSEMBLY_TIMEOUT),
+                            0,
+                        ));
+                        check!(fragments.get_packet_assembler_mut(&key))
+                    }
+                };
+
+                if !ipv4_packet.more_frags() {
+                    // This is the last fragment, so we know the total size
+                    check!(f.set_total_size(
+                        ipv4_packet.total_len() as usize - ipv4_packet.header_len() as usize
+                            + ipv4_packet.frag_offset() as usize,
+                    ));
+                }
+
+                match f.add(
+                    ipv4_packet.payload(),
+                    ipv4_packet.frag_offset() as usize,
+                    self.now,
+                ) {
+                    Ok(true) => {
+                        if !ipv4_packet.more_frags() {
```

Comment:
You shouldn't do this check here. If fragments arrive out of order, the last one to arrive might not be the last one in order, so it might have `more_frags = true`.

Reply:
Indeed, removed it.